File: geocode_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
地址转经纬度Python测试脚本
用于测试OpenStreetMap的Nominatim API
"""

import requests
import time
import json
import urllib.parse
from typing import Dict, List, Optional, Tuple, Union

# 地理编码缓存
geocode_cache = {}

class GeocodeError(Exception):
    """地理编码错误异常"""
    pass

def address_to_coordinates(address: str) -> Dict[str, Union[str, float]]:
    """
    使用OpenStreetMap的Nominatim服务将地址转换为经纬度坐标
    
    Args:
        address (str): 要转换的地址
    
    Returns:
        Dict: 包含lat（纬度）、lng（经度）和address（规范化地址）的字典
    
    Raises:
        GeocodeError: 地理编码失败时抛出
    """
    # 检查缓存
    if address in geocode_cache:
        print(f"[缓存命中] {address}")
        return geocode_cache[address]
    
    # 处理空地址
    if not address or address.strip() == '':
        raise GeocodeError("地址不能为空")
    
    try:
        # 编码地址
        encoded_address = urllib.parse.quote(address)
        
        # Nominatim API请求URL
        url = f"https://nominatim.openstreetmap.org/search?q={encoded_address}&format=json&limit=1&accept-language=zh-CN"
        
        print(f"[请求] {address}")
        
        # 设置User-Agent以符合Nominatim使用政策
        headers = {
            'User-Agent': 'wm_map_app/1.0 (employee map application)',
            'Accept': 'application/json'
        }
        
        # 发送请求
        response = requests.get(url, headers=headers, timeout=10)
        
        if not response.ok:
            raise GeocodeError(f"HTTP错误: {response.status_code} - {response.text}")
        
        # 解析JSON响应
        data = response.json()
        
        if data and len(data) > 0:
            location = {
                'lat': float(data[0]['lat']),
                'lng': float(data[0]['lon']),
                'address': data[0].get('display_name', address)
            }
            
            # 存入缓存
            geocode_cache[address] = location
            
            print(f"[成功] {address} -> ({location['lat']}, {location['lng']})")
            return location
        else:
            error_msg = f"未找到地址: {address}"
            print(f"[失败] {error_msg}")
            raise GeocodeError(error_msg)
    
    except requests.exceptions.RequestException as e:
        error_msg = f"网络请求错误: {str(e)}"
        print(f"[错误] {error_msg}")
        raise GeocodeError(error_msg)
    except (ValueError, KeyError) as e:
        error_msg = f"数据解析错误: {str(e)}"
        print(f"[错误] {error_msg}")
        raise GeocodeError(error_msg)

def batch_address_to_coordinates(addresses: List[str], delay_ms: int = 1000) -> List[Dict]:
    """
    批量转换地址为坐标
    
    Args:
        addresses (List[str]): 地址列表
        delay_ms (int): 请求间隔时间（毫秒），默认1000ms
    
    Returns:
        List[Dict]: 包含每个地址转换结果的列表
    """
    results = []
    delay_seconds = delay_ms / 1000.0
    
    print(f"\n开始批量处理 {len(addresses)} 个地址，请求间隔 {delay_ms}ms...")
    
    for i, address in enumerate(addresses):
        try:
            # 转换地址
            coordinates = address_to_coordinates(address)
            results.append({
                'address': address,
                'coordinates': coordinates,
                'success': True
            })
        except GeocodeError as e:
            results.append({
                'address': address,
                'error': str(e),
                'success': False
            })
        
        # 添加延迟避免触发速率限制
        if i < len(addresses) - 1:
            print(f"[等待] 下一个请求前等待 {delay_seconds} 秒...")
            time.sleep(delay_seconds)
    
    return results

def print_batch_results(results: List[Dict]) -> None:
    """
    打印批量处理结果
    
    Args:
        results (List[Dict]): 批量处理结果列表
    """
    success_count = sum(1 for r in results if r['success'])
    error_count = len(results) - success_count
    
    print(f"\n===== 批量处理结果摘要 =====")
    print(f"总地址数: {len(results)}")
    print(f"成功: {success_count}")
    print(f"失败: {error_count}")
    print("==========================")
    
    # 打印详细结果
    print("\n详细结果:")
    for i, result in enumerate(results):
        print(f"\n地址 {i+1}:")
        print(f"  原始地址: {result['address']}")
        
        if result['success']:
            coords = result['coordinates']
            print(f"  规范化地址: {coords['address']}")
            print(f"  纬度 (lat): {coords['lat']}")
            print(f"  经度 (lng): {coords['lng']}")
            print(f"  坐标对: [{coords['lat']}, {coords['lng']}]")
        else:
            print(f"  状态: 失败")
            print(f"  错误: {result['error']}")

def save_results_to_json(results: List[Dict], filename: str = 'geocode_results.json') -> None:
    """
    将结果保存到JSON文件
    
    Args:
        results (List[Dict]): 要保存的结果列表
        filename (str): 输出文件名
    """
    try:
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(results, f, ensure_ascii=False, indent=2)
        print(f"\n结果已保存到: {filename}")
    except Exception as e:
        print(f"\n保存结果失败: {str(e)}")

def run_interactive_test():
    """
    交互式测试函数
    """
    print("===== OpenStreetMap Nominatim 地址转经纬度测试工具 =====")
    print("使用说明:")
    print("1. 输入 '1' 测试单个地址")
    print("2. 输入 '2' 测试批量地址")
    print("3. 输入 '3' 查看缓存信息")
    print("4. 输入 '4' 清除缓存")
    print("5. 输入 'exit' 退出程序")
    print("====================================================")
    
    while True:
        choice = input("\n请选择操作: ").strip().lower()
        
        if choice == 'exit':
            print("谢谢使用，再见！")
            break
        
        elif choice == '1':
            # 单个地址测试
            address = input("请输入要转换的地址: ").strip()
            if not address:
                print("地址不能为空！")
                continue
            
            try:
                print(f"\n正在转换: '{address}'")
                result = address_to_coordinates(address)
                print("\n转换结果:")
                print(f"  原始地址: {address}")
                print(f"  规范化地址: {result['address']}")
                print(f"  纬度 (lat): {result['lat']}")
                print(f"  经度 (lng): {result['lng']}")
                print(f"  坐标对: [{result['lat']}, {result['lng']}]")
            except GeocodeError as e:
                print(f"\n转换失败: {str(e)}")
        
        elif choice == '2':
            # 批量地址测试
            print("请输入多个地址，每行一个（输入空行结束）:")
            addresses = []
            while True:
                line = input().strip()
                if not line:
                    break
                addresses.append(line)
            
            if not addresses:
                print("没有输入地址！")
                continue
            
            delay = input("请输入请求间隔时间(毫秒，默认为1000ms): ").strip()
            delay_ms = int(delay) if delay and delay.isdigit() else 1000
            
            results = batch_address_to_coordinates(addresses, delay_ms)
            print_batch_results(results)
            
            save_choice = input("\n是否保存结果到JSON文件? (y/n): ").strip().lower()
            if save_choice == 'y':
                filename = input("请输入文件名(默认为'geocode_results.json'): ").strip()
                if not filename:
                    filename = 'geocode_results.json'
                save_results_to_json(results, filename)
        
        elif choice == '3':
            # 查看缓存信息
            print(f"\n当前缓存信息:")
            print(f"缓存地址数量: {len(geocode_cache)}")
            if geocode_cache:
                print("缓存内容:")
                for address, location in geocode_cache.items():
                    print(f"  - '{address}' -> ({location['lat']}, {location['lng']})")
        
        elif choice == '4':
            # 清除缓存
            geocode_cache.clear()
            print("\n缓存已清除")
        
        else:
            print("无效的选择，请重新输入！")

def run_demo_test():
    """
    运行演示测试
    """
    print("\n===== 运行演示测试 =====")
    
    # 测试单个地址
    print("\n1. 单个地址测试:")
    demo_addresses = [
        "北京市海淀区中关村大街1号",
        "上海市浦东新区世纪大道1号"
    ]
    
    for address in demo_addresses:
        try:
            result = address_to_coordinates(address)
            print(f"  地址: {address}")
            print(f"  坐标: [{result['lat']}, {result['lng']}]")
        except GeocodeError as e:
            print(f"  地址: {address} - 失败: {str(e)}")
    
    # 测试批量地址
    print("\n2. 批量地址测试:")
    batch_addresses = [
        "广州市天河区天河路",
        "深圳市南山区科技园",
        "杭州市西湖区文三路",
        "不存在的地址123456789"
    ]
    
    results = batch_address_to_coordinates(batch_addresses)
    print_batch_results(results)

if __name__ == "__main__":
    # 首先运行演示测试
    run_demo_test()
    
    # 然后启动交互式测试
    run_interactive_test()

File: geocode_test.js
// 独立的地址转经纬度测试脚本
// 这个脚本可以单独运行，也可以在浏览器中通过简单的HTML调用

// 地理编码缓存，避免重复调用API
const geocodeCache = new Map();

/**
 * 使用OpenStreetMap的Nominatim服务将地址转换为经纬度坐标
 * @param {string} address - 要转换的地址
 * @returns {Promise<Object>} - 包含lat和lng的坐标对象
 */
async function addressToCoordinates(address) {
    // 检查缓存
    if (geocodeCache.has(address)) {
        console.log('使用缓存的地理编码结果:', address);
        return geocodeCache.get(address);
    }
    
    // 处理空地址
    if (!address || address.trim() === '') {
        throw new Error('地址不能为空');
    }
    
    try {
        // 编码地址
        const encodedAddress = encodeURIComponent(address);
        
        // Nominatim API请求URL
        const url = `https://nominatim.openstreetmap.org/search?q=${encodedAddress}&format=json&limit=1&accept-language=zh-CN`;
        
        console.log('正在请求地理编码:', address);
        
        // 发送请求，设置User-Agent以符合Nominatim使用政策
        const response = await fetch(url, {
            headers: {
                'User-Agent': 'wm_map_app/1.0 (employee map application)'
            }
        });
        
        if (!response.ok) {
            throw new Error(`地理编码请求失败: ${response.status}`);
        }
        
        const data = await response.json();
        
        if (data && data.length > 0) {
            const location = {
                lat: parseFloat(data[0].lat),
                lng: parseFloat(data[0].lon),
                address: data[0].display_name || address
            };
            
            // 存入缓存
            geocodeCache.set(address, location);
            
            console.log('地理编码成功:', address, location);
            return location;
        } else {
            console.warn('未找到地址:', address);
            throw new Error(`未找到地址: ${address}`);
        }
    } catch (error) {
        console.error('地理编码出错:', error);
        throw error;
    }
}

/**
 * 批量转换地址为坐标
 * @param {Array<string>} addresses - 地址数组
 * @param {number} delayMs - 请求间隔时间（毫秒），默认1000ms
 * @returns {Promise<Array<Object>>} - 包含地址和坐标的结果数组
 */
async function batchAddressToCoordinates(addresses, delayMs = 1000) {
    const results = [];
    
    for (let i = 0; i < addresses.length; i++) {
        try {
            const address = addresses[i];
            const coordinates = await addressToCoordinates(address);
            results.push({
                address,
                coordinates,
                success: true
            });
        } catch (error) {
            results.push({
                address: addresses[i],
                error: error.message,
                success: false
            });
        }
        
        // 添加延迟避免触发速率限制
        if (i < addresses.length - 1) {
            await new Promise(resolve => setTimeout(resolve, delayMs));
        }
    }
    
    return results;
}

// 如果在浏览器环境中运行，添加简单的测试界面支持
if (typeof window !== 'undefined' && typeof document !== 'undefined') {
    // 等待DOM加载完成
    document.addEventListener('DOMContentLoaded', function() {
        // 检查是否存在测试容器，如果没有则创建一个简单的测试界面
        let testContainer = document.getElementById('geocode-test-container');
        
        if (!testContainer) {
            // 创建简单的测试界面
            testContainer = document.createElement('div');
            testContainer.id = 'geocode-test-container';
            testContainer.style.maxWidth = '600px';
            testContainer.style.margin = '0 auto';
            testContainer.style.padding = '20px';
            testContainer.style.fontFamily = 'Arial, sans-serif';
            
            testContainer.innerHTML = `
                <h2>地址转经纬度测试工具</h2>
                
                <!-- 单个地址测试 -->
                <div style="margin-bottom: 30px;">
                    <h3>单个地址测试</h3>
                    <input type="text" id="single-address" placeholder="请输入地址" style="width: 100%; padding: 8px; margin-bottom: 10px;">
                    <button id="convert-btn" style="padding: 8px 16px; background-color: #4285f4; color: white; border: none; cursor: pointer;">
                        转换
                    </button>
                    <div id="single-result" style="margin-top: 15px; padding: 10px; border: 1px solid #ddd; min-height: 100px;">
                        结果将显示在这里
                    </div>
                </div>
                
                <!-- 批量地址测试 -->
                <div>
                    <h3>批量地址测试</h3>
                    <textarea id="batch-addresses" placeholder="请输入多个地址，每行一个" rows="5" style="width: 100%; padding: 8px; margin-bottom: 10px;"></textarea>
                    <button id="batch-convert-btn" style="padding: 8px 16px; background-color: #34a853; color: white; border: none; cursor: pointer;">
                        批量转换
                    </button>
                    <div id="batch-result" style="margin-top: 15px; padding: 10px; border: 1px solid #ddd; min-height: 100px; max-height: 300px; overflow-y: auto;">
                        结果将显示在这里
                    </div>
                </div>
            `;
            
            document.body.appendChild(testContainer);
        }
        
        // 绑定单个地址转换事件
        document.getElementById('convert-btn').addEventListener('click', async function() {
            const address = document.getElementById('single-address').value;
            const resultDiv = document.getElementById('single-result');
            
            if (!address) {
                resultDiv.innerHTML = '<span style="color: red;">请输入地址</span>';
                return;
            }
            
            resultDiv.innerHTML = '正在转换中...';
            
            try {
                const result = await addressToCoordinates(address);
                resultDiv.innerHTML = `
                    <p><strong>地址:</strong> ${result.address}</p>
                    <p><strong>纬度 (lat):</strong> ${result.lat}</p>
                    <p><strong>经度 (lng):</strong> ${result.lng}</p>
                    <p><strong>坐标:</strong> [${result.lat}, ${result.lng}]</p>
                `;
            } catch (error) {
                resultDiv.innerHTML = `<span style="color: red;">错误: ${error.message}</span>`;
            }
        });
        
        // 绑定批量地址转换事件
        document.getElementById('batch-convert-btn').addEventListener('click', async function() {
            const addressesText = document.getElementById('batch-addresses').value;
            const resultDiv = document.getElementById('batch-result');
            
            if (!addressesText) {
                resultDiv.innerHTML = '<span style="color: red;">请输入地址</span>';
                return;
            }
            
            const addresses = addressesText.split('\n')
                .map(line => line.trim())
                .filter(line => line !== '');
            
            resultDiv.innerHTML = `正在转换 ${addresses.length} 个地址，请稍候...`;
            
            try {
                const results = await batchAddressToCoordinates(addresses);
                
                // 显示结果
                let html = '<h4>转换结果:</h4>';
                results.forEach((result, index) => {
                    if (result.success) {
                        html += `
                            <div style="margin-bottom: 15px; padding: 10px; background-color: #f1f8e9;">
                                <p><strong>地址:</strong> ${result.address}</p>
                                <p><strong>纬度:</strong> ${result.coordinates.lat}</p>
                                <p><strong>经度:</strong> ${result.coordinates.lng}</p>
                            </div>
                        `;
                    } else {
                        html += `
                            <div style="margin-bottom: 15px; padding: 10px; background-color: #ffebee;">
                                <p><strong>地址:</strong> ${result.address}</p>
                                <p><strong>错误:</strong> ${result.error}</p>
                            </div>
                        `;
                    }
                });
                
                resultDiv.innerHTML = html;
            } catch (error) {
                resultDiv.innerHTML = `<span style="color: red;">错误: ${error.message}</span>`;
            }
        });
        
        // 添加键盘事件
        document.getElementById('single-address').addEventListener('keypress', function(e) {
            if (e.key === 'Enter') {
                document.getElementById('convert-btn').click();
            }
        });
    });
}

// 导出函数以便其他脚本使用
if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
    module.exports = {
        addressToCoordinates,
        batchAddressToCoordinates
    };
}

// 简单的测试示例（如果直接运行此脚本）
if (typeof window === 'undefined') {
    // 在Node.js环境中运行时的测试
    console.log('这是一个地址转经纬度的独立测试脚本');
    console.log('使用方法:');
    console.log('1. 浏览器环境: 引入此脚本到HTML页面');
    console.log('2. Node.js环境: const { addressToCoordinates } = require(\'./geocode_test.js\');');
    console.log('3. 直接调用: await addressToCoordinates(\'北京市海淀区\')');
}